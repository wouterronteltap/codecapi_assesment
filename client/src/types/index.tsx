import {Dispatch} from "react";

// TODO REWRITE PER ACTION
export interface IAction {
    type: string;
    // TODO => PROPER TYPE FOR PAYLOAD
    payload?: any;
}

export interface ILinks {
    last?: string,
    next?: string
}

export type IUser = {
    login: string,
    id: number,
    score?: number,
    avatar_url?: string,
    company?: string,
    email?: string,
    followers: number,
    following: number,
    name?: string,
    location?: string,
    bio?: string,
    blog?: string,
    repos_url: string,
    html_url: string

}

export interface IUsersResponse {
    users:  Array<IUser>,
    links: ILinks,
    total: number
}

export interface IState<T> {
    query: string,
    currentPage: number,
    total: number,
    lastPage: number,
    isLoading: boolean;
    isError: boolean;
    data?: T;
}

export interface IDispatch {
    dispatch: Dispatch<IAction>;
}
