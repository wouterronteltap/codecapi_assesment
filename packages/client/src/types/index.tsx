import { Dispatch } from "react";
import { AxiosResponse, AxiosError } from "axios";

type Actions =
  | "FETCH_INIT"
  | "FETCH_SUCCESS"
  | "FETCH_FAILURE"
  | "SET_PAGE"
  | "SET_DEBOUNCED_QUERY";

export interface IAction {
  type: Actions;
  payload: any;
}

export interface ILinks {
  last?: string;
  next?: string;
}

export type IUser = {
  login: string;
  id: number;
  score?: number;
  avatar_url?: string;
  company?: string;
  email?: string;
  followers: number;
  following: number;
  name?: string;
  location?: string;
  bio?: string;
  repos_url: string;
  html_url: string;
};

export interface IResponse extends AxiosResponse {
  users: Array<IUser>;
  links?: ILinks;
  total?: number;
}

export interface IState {
  query: string;
  currentPage: number;
  total: number;
  lastPage: number;
  isLoading: boolean;
  isError: boolean;
  data: IResponse | null;
  error: AxiosError | null;
}

export interface IDispatch {
  dispatch: Dispatch<IAction>;
}
