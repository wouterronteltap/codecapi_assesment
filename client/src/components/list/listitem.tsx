import React from 'react';
import { ListItem, ListItemText, ListItemAvatar, Avatar, ListItemSecondaryAction, IconButton, Typography } from "@material-ui/core";
import { Link } from 'react-router-dom';
import { IUser } from "../../types";

interface IProps {
    user: IUser
}

const CustomListItem = (props : IProps) => {
    const {user} = props;
    return(
        <ListItem>
            <ListItemAvatar>
                <Avatar src={user.avatar_url}/>
            </ListItemAvatar>
            <ListItemText primary={user.login} secondary={'Score: '+user.score}/>
            <ListItemSecondaryAction>
                <Link to={'/user/'+user.login}>
                    more
                </Link>
            </ListItemSecondaryAction>
        </ListItem>
    )
}

export { CustomListItem };