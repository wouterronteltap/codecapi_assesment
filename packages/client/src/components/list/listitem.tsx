import React from "react";
import {
  ListItem,
  ListItemText,
  ListItemAvatar,
  Avatar,
  ListItemSecondaryAction
} from "@material-ui/core";
import { Link } from "react-router-dom";
import { IUser } from "../../types";

interface IProps {
  user: IUser;
}

const CustomListItem = (props: IProps) => {
  const { user } = props;
  return (
    <ListItem>
      <ListItemAvatar>
        <Avatar src={user.avatar_url} />
      </ListItemAvatar>
      <ListItemText primary={user.login} secondary={"Score: " + user.score} />
      <ListItemSecondaryAction>
        <Link className={"showmore"} to={"/user/" + user.login}>
          more
        </Link>
      </ListItemSecondaryAction>
    </ListItem>
  );
};

export { CustomListItem };
