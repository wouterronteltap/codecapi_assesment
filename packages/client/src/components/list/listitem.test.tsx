import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { CustomListItem } from "./listitem";
import { IUser } from "../../types";

const userMock: IUser = {
  id: 1,
  name: "Wouter",
  score: 9,
  login: "Woutertje",
  avatar_url: "fakepath.png",
  followers: 19,
  following: 10,
  repos_url: "/fakepath/repos",
  html_url: "/fakepath"
};

describe("The CustomListItem Component", () => {
  afterEach(cleanup);
  test("renders correctly", () => {
    const { asFragment } = render(<CustomListItem user={userMock} />, {
      wrapper: MemoryRouter
    });
    expect(asFragment()).toMatchSnapshot();
  });

  test("should have link and navigate", () => {
    const { getByText } = render(<CustomListItem user={userMock} />, {
      wrapper: MemoryRouter
    });
    fireEvent.click(getByText("more"));
    console.log(window.location.pathname);
  });
});
