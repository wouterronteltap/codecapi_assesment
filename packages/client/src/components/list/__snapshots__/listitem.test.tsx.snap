// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The CustomListItem Component renders correctly 1`] = `
<DocumentFragment>
  <li
    class="MuiListItem-container"
  >
    <div
      class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
    >
      <div
        class="MuiListItemAvatar-root"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle"
        >
          <img
            class="MuiAvatar-img"
            src="fakepath.png"
          />
        </div>
      </div>
      <div
        class="MuiListItemText-root MuiListItemText-multiline"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
        >
          Woutertje
        </span>
        <p
          class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          Score: 9
        </p>
      </div>
    </div>
    <div
      class="MuiListItemSecondaryAction-root"
    >
      <a
        class="showmore"
        href="/user/Woutertje"
      >
        more
      </a>
    </div>
  </li>
</DocumentFragment>
`;
